### PR TITLE
[GsOC] Implementation of Zippel's GCD algorithm

### PR DESCRIPTION
**Pull request by luca-berton1**
#### Brief description of what is fixed or changed
Added Zippel's GCD algorithm for both the monic and not monic case.

#### Other comments
Apart of the 3 main functions implementing Zippel's GCD algorithm, this PR implements various helper functions, used by the main functions.

#### AI Generation Disclosure

Code was written by myself. I used AI to get reviews and help in writing the docstrings.

#### Release Notes

* polys
  * Added Zippel's GCD algorithm for sparse polynomials.

### CODE REVIEW

**Comment by luca-berton1** [comment]
I will soon add various GCD tests to this PR in a separate commit, to ensure that the functions are solid on all the test cases.

**Comment by luca-berton1** [comment]
Hello @oscarbenjamin, I give you a list of references for the algorithm:

- K. O. Geddes, S. R. Czapor, G. Labahn. Algorithms for Computer Algebra. p. 300-335, 1992.

This book provides a worked example of Zippel's algorithm, it's example 7.16, it's not very specific and therefore not super useful for the actual implementation, but I would recommend reading this example before everything else, to understand the algorithm.

- Computing the Greatest Common Divisor of
Multivariate Polynomials over Finite Fields, Suling Yang

This is the main reference I used, since it provides pseudocode for the algorithm.

Actually the paper talks about an adaptation of Zippel's algorithm for GCD's over small finite fields, but the logic behind it is analogous to the logic behind the classic Zippel algorithm: instead of Z here we have Z_p, and instead of quotienting over p in this paper the quotient is done over an irreducible polynomial: if one reads the paper knowing this and mentally translates `Z_p -> Z` and irreducible `poly -> p`, then the paper becomes a good resource.
It also describes in detail how the sparse interpolation using vandermonde matrices is performed (section 2.4), and how the not monic case is handled (section 2.5).
The functions I implemented follow the pseudocode provided in figures 3,4,5,6:
-figure 3 describes `modgcd_multivariate`
-figure 4 describes `sparse_gcd`
-figures 5/6 describe `zippel_interp`
 I decided not to write 2 distinct functions for the monic and non monic case since it would've resulted in a lot of code    duplication.

-Algorithms for the Non-monic Case of the Sparse Modular GCD Algorithm, De Kleine , Monagan, Wittkopf

This is useful since it provides examples of the `normalization` problem (example 1), and the `content` problem (example 4).

-My GSOC proposal:  https://docs.google.com/document/d/1xkyiVnZ3skfN-1WbIgdJFAWNxqLbUghY53K9AZT6loA/edit?usp=sharing

Here I tried to describe the math behind the algorithm, and also the concrete implementation. I think it could be useful. 
-section 1.1 describes the math
-from page 17 (Section: The concrete implementation in SymPy) there is a description of what the functions do, (of course since I wrote this before writing the actual functions it doesn't go into full detail).
Some functions described in the section have different names in this PR:
`sparse_gcd` is `modgcd_multivariate`,
 `_sparse_modgcd` is `sparse_gcd`,
 `zippel_interp_monic/zippel_interp_not_monic` are both implemented in `zippel_interp`,   `dense_incremental_interpolation` is implemented by `incremental_newton_interp` and `from_newt_to_poly`,
 and `vandermonde_solver` is implemented by `lag_basis` and `vandermonde_interp`.

For any questions I'm available of course.

**Comment by luca-berton1** [comment]
Hi Oscar, I just wrote some posts on my GSoC blog, with the help of chatGPT, describing in detail this PR, and also Zippel's algorithm in general. 

If you haven't looked at the sources above yet then I think that those last posts could be useful, because they get into mathematical detail, and they describe the steps of the concrete implementation. 

The post "[The Interpolation Machinery Behind Zippel's Algorithm](https://luca-berton1.github.io/the-interpolation-machinery-behind-zippels-algorithm/)" explains a bit what I think is the crucial node of the algorithm, that is passing from n-1 to n variables, and explains how this is done using the helper functions of the open PRs. 

The post "[Implementing Zippel's Algorithm: Architecture of the Monic Case](https://luca-berton1.github.io/Zippel's-algorithm-the-monic-case/)" explains this PR: I haven't added pieces of code into the posts (also because the PR will be changed after the refactoring), but the post follows the logic structure of the code. 

For the moment in the blog I haven't described the non-monic case yet, I will soon write something about how it is handled.

**Comment by luca-berton1** [comment]
Hello @oscarbenjamin ,I just finished adding tests to this PR.
Some of them are written by myself, while others were written by the AI. My workflow for this second cathegory was to ask the AI if it could write a test triggering a certain scenario in the algorithm that I wanted to test, and then to check if it actually triggered the desired scenario.
Now the PR has 100% test coverage. It took me longer than I thought because the tests I wrote highlighted some edge cases that weren't handled well in the original algorithm, and so I also had to modify the algorithm itself, and also because it wasn't always obvious which kind of input could trigger a certain scenario.
Now I will go on with the hypothesis testing.

**Comment by luca-berton1** [comment]
Hi @oscarbenjamin, I just added the hypothesis tests, and now this is ready for review.
I added 5 different types of tests: one for not coprime inputs, the second for most likely coprime inputs, and then also 3 other tests:
one for monic polynomials, one for pseudomonic polynomials, and one for not monic polynomials. 
These tests take different paths of the algorithm, and so I thought it would make sense to ensure that all paths are tested.

**Comment by oscarbenjamin** [comment]
Okay I'll take a look

File: sympy/polys/tests/test_zippel.py [hunk 1]
```diff
@@ -114,12 +141,429 @@ def test_smp_chinese_remainder_reconstruction_multivariate():
...
+    H1 = 1 + 2 * y + 3 * y**2 + 4 * y**4
+    H0 = 1 + 3 * y + 5 * y**2 + 7 * y**4
+    h = x**2 * (H2 + (z - 1) * y**3) + x * (H1 + (z - 1) * y**5) + H0 + (z - 1) * y**6
+    f = (x + 1) * h
+    g = (x + 2) * h
+    with patch(
+        "sympy.polys.zippel._random_eval_point",
+        side_effect=[1, 17, 23, 2, 3, 4, 5, 6, 7, 8],
+    ):
+        gcd, cff, cfg = smp_zippel_gcd(f, g, 3)
```
Comment (oscarbenjamin) [inline sympy/polys/tests/test_zippel.py]: What is the reason for using `patch` here?

I think it would be best to avoid using it.

Comment (luca-berton1) [reply sympy/polys/tests/test_zippel.py]: It is that zippel's algorithm relies on the fact that it picks random numbers, and with a small probability these random numbers will make the algorithm fail at some point. Therefore, there are a lot of fallbacks, that have to detect if the algorithm has run into one of those unlucky choices. Since the probability of having an unlucky choice is very small though, without using patch to make the random choices deterministic, it would become impossible to trigger the fallbacks of the algorithm.

Comment (luca-berton1) [reply sympy/polys/tests/test_zippel.py]: For example consider this test:
```
# wrong gcd is determined by recursive call due to unlucky evaluation of z
    R, x, y, z = ring("x, y, z", ZZ)
    f = (x + y - z) * (x**2 * y + x * (z - 3) + y * (z - 2) + (z - 5))
    g = (x + y + z) * (x**2 * y + x * (z - 3) + y * (z - 2) + (z - 5))
    with patch(
        "sympy.polys.zippel._random_eval_point",
        side_effect=[2, 3, 5],
    ):
        gcd = smp_zippel_gcd_mod(f, g, ZZ(10009), 3)
    assert gcd is None
```

What Zippel algorithm does here, is that `smp_zippel_gcd` projects the polys coefficients over `Z_p,` and then `smp_zippel_gcd_mod` evaluates the last variable in a point `p` , and calls itself recursively on the evaluated polynomials. 

The recursive call is used to determine the gcd in the first 2 variables: once that gcd is returned, the algorithm relies on the assumption, that the gcd in 2 variables, will have all the monomials of the final gcd, in the first 2 variables.

So for example in the case of the test, the algorithm assumes that the recursive call returns something like this:
`x**2 * y + x * c1 + y * c2 + c3`
With `c1 = p - 3 , c2 = p - 2, c5 = p - 5` . So each of the coefficients is a polynomial in the variable `z`, evaluated in `p`. (Of course it could be the constant polynomial, as is the case for the coefficient of `x**2 * y` )

In the case of the test, by forcing `p = 2`, the recursive call to `smp_zippel_gcd_mod` returns:
`x**2 * y + x * c1  + c3`

This error will lead then the algorithm to not be able to compute the gcd correctly, and the fallback path will be activated.

The thing is, that here the prime used is in the order of `10**9` , and only the choices of `2,3 and 5` as `p` would have led to this kind of error, so I think that without somehow being able to deterministically simulate the random choices of the algorithm, it would be impossible to test this.

Comment (oscarbenjamin) [reply sympy/polys/tests/test_zippel.py]: Maybe `smp_zippel_gcd_mod` could have an optional parameter to pass in a function that computes the evaluation points.

Comment (luca-berton1) [reply sympy/polys/tests/test_zippel.py]: I modified the tests using your suggestion. In order to maintain the tests I had to create more such functions for the different casual extractions that the algorithm makes.

**Comment by luca-berton1** [comment]
Hello, previously one of the hypothesis tests activated the 10 seconds timeout, so I reduced the cases of every test to 25, by using `@settings(max_examples= 25)`

File: sympy/polys/zippel.py [hunk 2]
```diff
@@ -186,15 +223,908 @@ def crt_scalar(cp, cq, p, q):
...
+def _random_eval_point(p: MPZ) -> MPZ:
+    return ZZ(randint(1, int(p - 1)))
+
+
+def _solve_scaling_system(
+    A_m: DomainMatrix, B_m: DomainMatrix
+) -> None | int | tuple[DomainMatrix, Any]:
+    ncols = A_m.shape[1]
+    aug = A_m.hstack(B_m)
+    aug_rref, den, pivots = aug.rref_den()
```
Comment (oscarbenjamin) [inline sympy/polys/zippel.py:480]: It would be worth testing if `rref()` is faster than `rref_den()` since this is over a field.

Also it would be worth chcking if using the sparse representation of DomainMatrix is faster.

File: sympy/polys/zippel.py [hunk 3]
```diff
@@ -186,15 +223,908 @@ def crt_scalar(cp, cq, p, q):
...
+        m *= p
+
+        if not hm == hlastm:
+            hlastm = hm
+            continue
+
+        _, h = smp_primitive(hm, n, ZZ)
+        fquo, frem = smp_div_list(f, [h], n, ZZ)
+        gquo, grem = smp_div_list(g, [h], n, ZZ)
+        if not frem and not grem:
```
Comment (oscarbenjamin) [inline sympy/polys/zippel.py:692]: I wonder if it might be faster to do the division check first rather than checking `hm == hlastm` because the division might be cheaper than another call to `smp_zippel_gcd_mod`.

File: sympy/polys/zippel.py [hunk 4]
```diff
@@ -186,15 +223,908 @@ def crt_scalar(cp, cq, p, q):
...
+    for i in range(n):
+        sf = smp_swap(f, i, n, ZZ)
+        sg = smp_swap(g, i, n, ZZ)
+        badprimes *= ZZ.gcd(smp_LC(sf, n, ZZ), smp_LC(sg, n, ZZ))
+
+    # empirical evidence shows that brown is faster using one
+    # while Zippel needs large primes to not run out of evaluation points
+    m = ZZ.one
+
+    p = ZZ(10**9)
```
Comment (oscarbenjamin) [inline sympy/polys/zippel.py:647]: I wonder if it is worth computing how many primes to use. I think maybe that the product of the primes needs to be only bigger than something like `2*B` where `B` is the largest absolute value coefficient in `f` and `g`.

Should the choice of the prime `p` here depend on the degree or the number of terms or something?

File: sympy/polys/zippel.py [hunk 5]
```diff
@@ -186,15 +223,908 @@ def crt_scalar(cp, cq, p, q):
...
+            continue
+
+        hp = smp_mul_ground(hp, gamma, n, ZZ)
+        hp = smp_trunc_ground(hp, p, n, ZZ)
+        if m == 1:
+            m = p
+            hlastm = hp
+            continue
+
+        hm = smp_chinese_remainder_reconstruction_multivariate(hp, hlastm, p, m, ZZ, n)
```
Comment (oscarbenjamin) [inline sympy/polys/zippel.py:682]: Maybe we can just call this `smp_crt`.

File: sympy/polys/zippel.py [hunk 6]
```diff
@@ -7,15 +7,41 @@
...
     smp_LC,
+    smp_is_one,
+    smp_mul,
+    smp_mul_ground,
+    smp_primitive,
     smp_quo_ground,
+    smp_rem_list,
+    smp_subs_drop,
     smp_trunc_ground,
 )
```
Comment (oscarbenjamin) [inline sympy/polys/zippel.py:47]: Maybe the `crt` code here should use `crt1` and `crt2`? I'm not sure if those are applicable to this.

**Comment by oscarbenjamin** [comment]
I left comments for future consideration but let's get this in now.

**Comment by oscarbenjamin** [comment]
Probably smp_zippel_gcd_mod and smp_zippel_interp_mod should be private (leading underscore in the name). Still though their docstrings should explain what preprocessing is needed and what consitutes valid arguments to the functions (same for `smp_zippel_gcd`).

**Comment by oscarbenjamin** [comment]
This is what GPT complained about when reviewing the PR:
```
smp_zippel_gcd_mod and smp_zippel_interp_mod can loop forever when no valid evaluation point/tuple is available. In sympy/polys/
    zippel.py:804, repeated t in M or gk == 0 just continues with no attempt limit. Similarly, sympy/polys/zippel.py:956 retries invalid tuples
    indefinitely. This is deterministic for small fields such as p=2, where _random_eval_point can only return 1. A direct call like
    smp_zippel_gcd_mod((x + y)*(x + 1), (x + y)*x, ZZ(2), 2) hangs instead of returning None. These loops should have bounded retries like the
    surrounding pt_chances/skeleton_chances logic.
```
I think it just means that the functions need to be documented better to clearly explain requirements on the inputs.

File: sympy/polys/zippel.py (hunk 4 above)
Comment (oscarbenjamin) [reply sympy/polys/zippel.py:647]: There should probably be a comment here explaining about `10**9`. Why not bigger? Why not smaller?
